comp/core/secrets/secretsimpl/secrets.go

```diff
@@ -299,6 +327,59 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
 	return finalConfig, nil
 }
 
+// whitelistHandles restricts what config settings may be updated
+// tests can override this to exercise functionality
+var whitelistHandles = []string{"api_key"}
+
+func (r *secretResolver) processSecretResponse(secretResponse map[string]string, useWhitelist bool) {
+	// notify subscriptions about the changes to secrets
+	for handle, secretValue := range secretResponse {
+		// if whitelist is enabled and the handle is not contained in it, skip it
+		if useWhitelist && whitelistHandles != nil && !slices.Contains(whitelistHandles, handle) {
+			continue
+		}
+		oldValue := r.cache[handle]
+		// if value hasn't changed, don't send notifications
+		if oldValue == secretValue {
+			continue
+		}
+		for _, secretCtx := range r.origin[handle] {
+			for _, sub := range r.subscriptions {
+				sub(handle, secretCtx.path, oldValue, secretValue)
+			}
+		}
+		// add results to the cache
+		r.cache[handle] = secretValue
+	}
+}
+
+// Refresh the secrets after they have been Resolved by fetching them from the backend again
+func (r *secretResolver) Refresh() error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	newHandles := make([]string, 0, len(r.cache))
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Do we actually need to clone it ?

Reply:
Yes, need to convert the map[string]... into a []string. Using `maps.Keys` for that, thanks for the tip!

Reply:
Ah yes I wrote that before realizing that it was a `map` :smile: